<a href="https://colab.research.google.com/github/LuckyStrike42/Analitica/blob/main/Reto_1_Modelo_Naive_Bayes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>




# **Reto 1 Naive Bayes**

0. Cargar las librerias de trabajo

In [ ]:
import numpy as np #funciones mat
import pandas as pd #comunicacion con excel
from sklearn.naive_bayes import GaussianNB #libreria Naive Bayes
from sklearn.metrics import confusion_matrix #matriz de confusion


1. Cargar la base de datos de excel.

In [ ]:
BD="/content/Longitudinal Survey.xlsx"
df=pd.read_excel(BD,sheet_name=0) #Traigo los datos
df=df.dropna() #Elimino los valores nulos

#seleccionamos las variables de interes
df=df[['Age','Mother_Edu','Father_Edu','Height','Weight','Income','Self_Esteem','FamilySize','Urban']]

#Datos de entrada
dx=np.array(df[['Age','Mother_Edu','Father_Edu','Height','Weight','Income','Self_Esteem','FamilySize']])

#Datos de salida - Valor deseado - Que quiero que aprenda el modelo
dy=np.array(df['Urban'])


# **Preguntas**


¿Cuál es el número de datos total de la base de datos?

In [ ]:
DATA= len(df) #Tamaño de la base de datos
print("Numero total de datos encontrados en el archivo:", DATA)

Numero total de datos encontrados en el archivo: 5683


¿Cuál es el porcentaje de los datos que hacen parte de cada una de las categorías de clasificación?

In [ ]:
#Personas que viven en la ciudad.
filas=np.where(dy==1)[0]
len(filas)
print("Porcentaje de las personas que viven en la ciudad:", len(filas)/len(df)*100,"%")

#Personas que viven en areas rurales.
filas=np.where(dy==0)[0]
len(filas)
print("Porcentaje de las personas que viven en las areas rurales:", len(filas)/len(df)*100,"%")

Porcentaje de las personas que viven en la ciudad: 79.35949322540911 %
Porcentaje de las personas que viven en las areas rurales: 20.640506774590886 %


# **Modelo**

In [ ]:
nmb=GaussianNB() #Modelo de Naive Bayes
nmb.fit(dx,dy) #Entreno el modelo
ydp=nmb.predict(dx) #Predigo los datos de entrada - Pero que el obtuvo

#digame las medias de las variables para cada categoria - PreApr/PreNeg
u=nmb.theta_ #Se estiman las medias
sigma=np.sqrt(nmb.var_) #Se estiman las varianzas

#Las variaziones de las variables
LI=u-sigma #Limite inferior
LS=u+sigma #Limite superior

print('La media de las variables por categoria:')
print(u)
print('Limite inferior:')
print(LI)
print('Limite superior:')
print(LS)

#Metrica de evaluacion - como se comporto el modelo indentificando PreApr/PreNeg
cm=confusion_matrix(dy,ydp)
print('Matriz de confusion:')
print(cm)



La media de las variables por categoria:
[[1.83810742e+01 1.07612958e+01 1.03324808e+01 6.72148338e+01
  1.48857630e+02 3.03672506e+04 2.20272805e+01 4.74168798e+00]
 [1.83569845e+01 1.11769401e+01 1.12565410e+01 6.69031042e+01
  1.44760976e+02 3.34278929e+04 2.26004435e+01 4.68780488e+00]]
Limite inferior:
[[ 1.58996876e+01  7.58892579e+00  6.46311334e+00  6.31369799e+01
   1.18746964e+02 -1.33465823e+03  1.78344685e+01  2.16621834e+00]
 [ 1.58811112e+01  7.79013864e+00  7.15097738e+00  6.26897933e+01
   1.14669654e+02 -2.69533531e+03  1.84154526e+01  2.24996381e+00]]
Limite superior:
[[2.08624608e+01 1.39336659e+01 1.42018483e+01 7.12926877e+01
  1.78968296e+02 6.20691595e+04 2.62200925e+01 7.31715762e+00]
 [2.08328577e+01 1.45637416e+01 1.53621047e+01 7.11164151e+01
  1.74852297e+02 6.95511211e+04 2.67854344e+01 7.12564594e+00]]
Matriz de confusion:
[[   2 1171]
 [   4 4506]]


**Intervalos de las variables**

* **Edad**
1.   Rural: [15-20]
2.   Urbana [15-20]

* **Altura**  
1. Rural: [63.1-71.2]
2. Urbana: [62.2-71.1]

* **Peso**  
1.  Rural: [118.7-178.9]
2.  Urbana: [114.6-174.8]


# **Evaluacion del modelo**




*   ¿Cuál es la exactitud alcanzada por el modelo?
*   ¿Cuál es la tasa de error alcanzada por el modelo?
*   ¿Cuál es la Sensibilidad alcanzada por el modelo?
*   ¿Cuál es la Especificidad alcanzada por el modelo?



In [ ]:
from os import pread
VN=cm[0,0] #Verdaderos Negativos
FP=cm[0,1] #Falsos Positivos
FN=cm[1,0] #Falsos Negativos
VP=cm[1,1] #Verdaderos Positivos

exactitud=(VP+VN)/(VP+VN+FP+FN) #Comportamiento general
tasa_error=(FP+FN)/(VP+VN+FP+FN) #Porcentaje de equivocaciones
sensibilidad=VP/(VP+FP) #Como se comporto frente a los positivos
especificidad=VN/(VN+FN) #Como se comporto frente alos negativos
precission=VP/(VP+FN) #Precicion para clasificar positivos.
PredNeg=VN/(VN+FP) #Precicion para clasificar negativos.

print('Exactitud:')
print(exactitud)
print('Tasa de error:')
print(tasa_error)
print('Sensibilidad:')
print(sensibilidad)
print('Especificidad:')
print(especificidad)
print('Precission:')
print(precission)
print('PredNeg:')
print(PredNeg)

Exactitud:
0.7932430054548654
Tasa de error:
0.2067569945451346
Sensibilidad:
0.793729082261758
Especificidad:
0.3333333333333333
Precission:
0.9991130820399113
PredNeg:
0.0017050298380221654


**De acuerdo con los resultados del codigo de la evaluacion del modelo:**

Exactitud del modelo: 79%

Tasa de error del modelo 20%

Sensibilidad del modelo: 79%

Especificidad del modelo: 33%

Precisión del modelo: 99%

Valor de prediccion de Negativos del modelo: 0.17%




Caso de un individuo


In [ ]:
Xun=np.array([[34,9,6,65,135,50000,30,3]]) #Persona del Excel.
yp=nmb.predict(Xun)

if yp==1:
  print("La persona vive en la ciudad")
else:
  print("La persona vive en las areas rurales")


La persona vive en la ciudad


# **Analisis de resultados**



El modelo Naive Bayes aplicado a los datos de la encuesta logró una
exactitud del 79.32%, lo que significa que pudo predecir correctamente la mayoría de los casos. Aun así, hubo un 20.68% de errores, especialmente en clasificaciones incorrectas de personas que viven en áreas rurales como urbanas. Aunque el modelo fue bastante bueno para identificar a quienes residen en la ciudad, con una sensibilidad del 79.37%, su capacidad para reconocer correctamente a quienes viven en áreas rurales fue bastante limitada, con una especificidad de solo el 33.33%. Esto indica una tendencia del modelo a etiquetar erróneamente a los residentes rurales como urbanos.

A pesar de estos desafíos, la precisión del modelo en clasificar correctamente a las personas como residentes urbanos fue muy alta, alcanzando un 99.91%. No obstante, cuando se trató de predecir la residencia en áreas rurales, el modelo tuvo fallos considerables, con un valor de predicción negativa de apenas 0.17%. En resumen, aunque el modelo tiene un buen rendimiento al identificar a los residentes urbanos, encuentra dificultades significativas para clasificar adecuadamente a los residentes rurales. Esto podría deberse a una desproporción en los datos o a una selección de características que no es capaz de diferenciar entre urbano y rural.